cbmc/proofs/mqtt_state.c

```diff
@@ -0,0 +1,556 @@
+  return pConn;
+}
+
+void ensure_IotMqttConnection_has_lists(IotMqttConnection_t pConn)
+{
+  // TODO: add code to make lists nondet nontrivial
+  // Consider using allocate_IotMqttSubscriptionList
+  IotListDouble_Create(&pConn->pendingProcessing);
+  IotListDouble_Create(&pConn->pendingResponse);
+  IotListDouble_Create(&pConn->subscriptionList);
+  return pConn;
+}
+
+bool valid_IotMqttConnection(IotMqttConnection_t pConn)
+{
+  if ( pConn == NULL ) return 0;
+
+  bool valid_references = pConn->references >= 1;
+
+  bool valid_pingreq =
+    ( valid_IotMqttOperation( &(pConn->pingreq) ) ) &&
+    ( pConn->pingreq.u.operation.type == IOT_MQTT_PINGREQ ) &&
+    ( pConn->pingreq.pMqttConnection == pConn ) &&
+    ( !pConn->pingreq.incomingPublish ) ;
+
+  return
+    valid_IotNetworkInterface( pConn->pNetworkInterface ) &&
+    valid_references &&
+    valid_pingreq;
+}
+
+/****************************************************************
+ * IotMqttNetworkInfo
+ ****************************************************************/
+
+IotMqttNetworkInfo_t *allocate_IotMqttNetworkInfo(IotMqttNetworkInfo_t *pInfo)
+{
+  if ( pInfo == NULL ) pInfo = malloc_can_fail( sizeof( *pInfo ) );
+  if ( pInfo == NULL ) return NULL;
+
+  if (pInfo->createNetworkConnection) {
+    // allocate setup member of union
+    pInfo->u.setup.pNetworkServerInfo = allocate_opaque_type();
+    pInfo->u.setup.pNetworkCredentialInfo = allocate_opaque_type();
+  } else {
+    // allocate network member of union
+    pInfo->u.pNetworkConnection = allocate_IotNetworkConnection();
+  }
+  pInfo->pNetworkInterface = allocate_IotNetworkInterface();
+  return pInfo;
+}
+
+bool valid_IotMqttNetworkInfo(IotMqttNetworkInfo_t *pInfo)
+{
+  if ( pInfo == NULL ) return 0;
+  return 1;
+}
+
+/****************************************************************
+ * IotMqttConnectInfo
+ ****************************************************************/
+
+IotMqttConnectInfo_t *allocate_IotMqttConnectInfo(IotMqttConnectInfo_t *pInfo)
+{
+  if ( pInfo == NULL ) pInfo = malloc_can_fail( sizeof( *pInfo ) );
+  if ( pInfo == NULL ) return NULL;
+
+  pInfo->pPreviousSubscriptions =
+    allocate_IotMqttSubscriptionArray(NULL,
+				     pInfo->previousSubscriptionCount);
+  pInfo->pWillInfo = allocate_IotMqttPublishInfo(NULL);
+  pInfo->pClientIdentifier = malloc_can_fail(pInfo->clientIdentifierLength);
+  pInfo->pUserName = malloc_can_fail(pInfo->userNameLength);
+  pInfo->pPassword = malloc_can_fail(pInfo->passwordLength);
+}
+
+bool valid_IotMqttConnectInfo(IotMqttConnectInfo_t *pInfo)
+{
+  return
+    pInfo &&
+
+    VALID_STRING(pInfo->pClientIdentifier, pInfo->clientIdentifierLength) &&
+    VALID_CBMC_SIZE( pInfo->clientIdentifierLength ) &&
+
+    VALID_STRING(pInfo->pUserName, pInfo->userNameLength) &&
+    VALID_CBMC_SIZE( pInfo->userNameLength ) &&
+
+    VALID_STRING(pInfo->pPassword, pInfo->passwordLength) &&
+    VALID_CBMC_SIZE( pInfo->passwordLength ) &&
+
+#ifdef SUBSCRIPTION_COUNT_MAX
```

Comment:
I wonder if we should have a prefix for proof related `#define`s

Reply:
We could do this in a later revision, but right now I prefer to focus on the proofs.